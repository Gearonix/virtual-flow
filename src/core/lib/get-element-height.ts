export interface GetElementHeightProps {
  entry?: ResizeObserverEntry
  element: Element
}

export const getElementHeight = ({ element, entry }: GetElementHeightProps) => {
  return (
    entry?.borderBoxSize[0].blockSize ?? element.getBoundingClientRect().height
  )
}
