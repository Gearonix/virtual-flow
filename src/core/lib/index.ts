export { calculateVirtualItems } from './calculate-virtual-items'
export { getElementHeight } from './get-element-height'
export { getCacheKey } from './get-item-cache-key'
