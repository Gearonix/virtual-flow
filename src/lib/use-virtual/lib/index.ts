export { calculateVirtualItems } from './calculate-virtual-items'
export { getElementHeight } from './get-element-height'
export { getMeasurementCacheByElement } from './get-measurement-cache-by-elem'
