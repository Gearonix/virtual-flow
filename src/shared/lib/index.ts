export { createDOMScheduler } from './dom-scheduler'
export { exhaustiveCheck } from './exhaustive-check'
export { rafThrottle } from './raf-throttle'
export { createTestingContextWrapper } from './testing-wrapper'
export { withPropsValidator } from './with-props-validator'
