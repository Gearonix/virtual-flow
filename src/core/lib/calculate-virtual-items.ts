import { ItemCacheKey } from '@/core/use-virtual.interfaces'
import { VirtualItem }  from '@/core/use-virtual.interfaces'

export interface CalculateVirtualItemsProps {
  rangeStart: number
  rangeEnd: number
  itemsCount: number
  overscan: number
  getItemKey: (idx: number) => ItemCacheKey
  getItemHeight: (idx: number) => number
}

export const calculateVirtualItems = ({
  rangeStart,
  itemsCount,
  rangeEnd,
  overscan,
  getItemKey,
  getItemHeight
}: CalculateVirtualItemsProps) => {
  let startIdx = -1
  let endIdx = -1

  const allRows: VirtualItem[] = new Array(itemsCount).fill(null)
  let totalHeight = 0

  for (let idx = 0; idx < allRows.length; idx++) {
    const itemKey = getItemKey(idx)

    const row: VirtualItem = {
      key: itemKey,
      idx,
      virtualHeight: getItemHeight(idx),
      offsetTop: totalHeight
    }

    totalHeight += row.virtualHeight
    allRows[idx] = row

    if (row.offsetTop + row.virtualHeight > rangeStart && !~startIdx) {
      startIdx = Math.max(0, idx - overscan)
    }

    if (row.offsetTop + row.virtualHeight >= rangeEnd && !~endIdx) {
      endIdx = Math.min(itemsCount - 1, idx + overscan)
    }
  }

  endIdx = endIdx === -1 ? itemsCount - 1 : endIdx

  const virtualItems = allRows.slice(startIdx, endIdx + 1)

  return {
    virtualItems,
    totalHeight
  }
}
