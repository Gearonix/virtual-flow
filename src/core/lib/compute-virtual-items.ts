import { ItemCacheKey } from '@/core/use-virtual.interfaces'
import { VirtualItem }  from '@/core/use-virtual.interfaces'

export interface CalculateVirtualItemsProps {
  rangeStart: number
  rangeEnd: number
  itemsCount: number
  overscan: number
  getItemKey: (idx: number) => ItemCacheKey
  getItemHeight: (idx: number) => number
}

/** calculateVirtualItems
 * @returns virtualItems {VirtualItem[]} - items to render depending on the scroll position
 * @returns totalHeight {number} - The total height of all elements
 */

export const computeVirtualItems = ({
  rangeStart,
  itemsCount,
  rangeEnd,
  overscan,
  getItemKey,
  getItemHeight
}: CalculateVirtualItemsProps) => {
  let startIdx = -1
  let endIdx = -1

  const allRows: VirtualItem[] = new Array(itemsCount).fill(null)
  let totalHeight = 0

  for (let idx = 0; idx < allRows.length; idx++) {
    const itemKey = getItemKey(idx)

    const row: VirtualItem = {
      key: itemKey,
      idx,
      virtualHeight: getItemHeight(idx),
      offsetTop: totalHeight
    }

    totalHeight += row.virtualHeight
    allRows[idx] = row

    if (!~startIdx && row.offsetTop + row.virtualHeight > rangeStart) {
      startIdx = Math.max(0, idx - overscan)
    }

    if (!~endIdx && row.offsetTop + row.virtualHeight >= rangeEnd) {
      endIdx = Math.min(itemsCount - 1, idx + overscan)
    }
  }

  endIdx = endIdx === -1 ? itemsCount - 1 : endIdx

  const virtualItems = allRows.slice(startIdx, endIdx + 1)

  return {
    virtualItems,
    totalHeight,
    allItems: allRows
  }
}
