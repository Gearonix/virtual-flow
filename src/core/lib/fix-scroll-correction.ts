import { MutableRefObject } from 'react'

import { LatestInstance }   from '@/core/use-virtual.interfaces'

export interface FixScrollCorrectionProps {
  height: number
  elementIdx: number
  latestInstance: MutableRefObject<LatestInstance>
}

export const fixScrollCorrection = ({
  height,
  elementIdx,
  latestInstance
}: FixScrollCorrectionProps) => {
  const { getScrollElement, scrollTop, allItems } = latestInstance.current

  const virtualElement = allItems[elementIdx]

  const delta = height - virtualElement.virtualHeight

  if (delta === 0 || virtualElement.offsetTop >= scrollTop) return

  const scrollElement = getScrollElement()

  if (scrollElement) {
    scrollElement.scrollBy(0, delta)
  }
}
