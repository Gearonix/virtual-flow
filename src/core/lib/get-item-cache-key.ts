import { MutableRefObject }        from 'react'

import { LatestInstance }          from '@/core/use-virtual.interfaces'
import { VIRTUAL_INDEX_ATTRIBUTE } from '@/shared/consts'
import { NoVirtualIndexException } from '@/shared/exceptions'
import { el } from '@faker-js/faker'

export interface UseMeasurementCacheByElementProps {
  element: Element
  latestInstance: MutableRefObject<LatestInstance>
}

/**
 * Returns the cache key of the element,
 * according to the data-vindex attribute.
 * @param element - DOM Element
 * @param latestInstance - instance of useLatest hook
 */

export const getCacheKey = ({
  element,
  latestInstance
}: UseMeasurementCacheByElementProps) => {
  const idxAttribute = element.getAttribute(VIRTUAL_INDEX_ATTRIBUTE) ?? ''
  const elementIdx = Number.parseInt(idxAttribute, 10)

  if (Number.isNaN(elementIdx)) {
    return
    // throw new NoVirtualIndexException()
  }

  const { getItemKey } = latestInstance.current

  return {
    cacheKey: getItemKey(elementIdx),
    elementIdx
  }
}
