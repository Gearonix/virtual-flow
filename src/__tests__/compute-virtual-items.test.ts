import { beforeEach }                 from 'vitest'
import { expect }                     from 'vitest'
import { describe }                   from 'vitest'
import { test }                       from 'vitest'

import { CalculateVirtualItemsProps } from '@/core/lib/compute-virtual-items'
import { computeVirtualItems }        from '@/core/lib/compute-virtual-items'

describe('compute-virtual-items', () => {
  let payload: CalculateVirtualItemsProps

  beforeEach(() => {
    payload = {
      rangeStart: 100,
      rangeEnd: 400,
      itemsCount: 1000,
      overscan: 5,
      getItemKey: (idx: number) => `item-${idx}`,
      getItemHeight: () => 50
    }
  })

  test('calculates virtual items and total height', () => {
    const result = computeVirtualItems(payload)

    expect(result.virtualItems).toHaveLength(13)
    expect(result.totalHeight).toBe(50_000)
  })

  test('handles case where rangeStart and rangeEnd are at the beginning', () => {
    payload.rangeStart = 0
    payload.rangeEnd = 1200

    const result = computeVirtualItems(payload)

    expect(result.virtualItems).toHaveLength(29)
    expect(result.totalHeight).toBe(50_000)
  })

  test('handles a small range with no overscan', () => {
    payload.rangeStart = 50
    payload.rangeEnd = 75

    const result = computeVirtualItems(payload)

    expect(result.virtualItems).toHaveLength(7)
    expect(result.totalHeight).toBe(50_000)
  })
})
