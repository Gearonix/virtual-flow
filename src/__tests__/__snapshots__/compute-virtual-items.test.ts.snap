// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compute-virtual-items > handles a small range with no overscan 1`] = `
[
  {
    "idx": 0,
    "key": "item-0",
    "offsetTop": 0,
    "virtualHeight": 50,
  },
  {
    "idx": 1,
    "key": "item-1",
    "offsetTop": 50,
    "virtualHeight": 50,
  },
  {
    "idx": 2,
    "key": "item-2",
    "offsetTop": 100,
    "virtualHeight": 50,
  },
  {
    "idx": 3,
    "key": "item-3",
    "offsetTop": 150,
    "virtualHeight": 50,
  },
  {
    "idx": 4,
    "key": "item-4",
    "offsetTop": 200,
    "virtualHeight": 50,
  },
  {
    "idx": 5,
    "key": "item-5",
    "offsetTop": 250,
    "virtualHeight": 50,
  },
  {
    "idx": 6,
    "key": "item-6",
    "offsetTop": 300,
    "virtualHeight": 50,
  },
]
`;

exports[`compute-virtual-items > handles case where rangeStart and rangeEnd are at the beginning 1`] = `
[
  {
    "idx": 0,
    "key": "item-0",
    "offsetTop": 0,
    "virtualHeight": 50,
  },
  {
    "idx": 1,
    "key": "item-1",
    "offsetTop": 50,
    "virtualHeight": 50,
  },
  {
    "idx": 2,
    "key": "item-2",
    "offsetTop": 100,
    "virtualHeight": 50,
  },
  {
    "idx": 3,
    "key": "item-3",
    "offsetTop": 150,
    "virtualHeight": 50,
  },
  {
    "idx": 4,
    "key": "item-4",
    "offsetTop": 200,
    "virtualHeight": 50,
  },
  {
    "idx": 5,
    "key": "item-5",
    "offsetTop": 250,
    "virtualHeight": 50,
  },
  {
    "idx": 6,
    "key": "item-6",
    "offsetTop": 300,
    "virtualHeight": 50,
  },
  {
    "idx": 7,
    "key": "item-7",
    "offsetTop": 350,
    "virtualHeight": 50,
  },
  {
    "idx": 8,
    "key": "item-8",
    "offsetTop": 400,
    "virtualHeight": 50,
  },
  {
    "idx": 9,
    "key": "item-9",
    "offsetTop": 450,
    "virtualHeight": 50,
  },
  {
    "idx": 10,
    "key": "item-10",
    "offsetTop": 500,
    "virtualHeight": 50,
  },
  {
    "idx": 11,
    "key": "item-11",
    "offsetTop": 550,
    "virtualHeight": 50,
  },
  {
    "idx": 12,
    "key": "item-12",
    "offsetTop": 600,
    "virtualHeight": 50,
  },
  {
    "idx": 13,
    "key": "item-13",
    "offsetTop": 650,
    "virtualHeight": 50,
  },
  {
    "idx": 14,
    "key": "item-14",
    "offsetTop": 700,
    "virtualHeight": 50,
  },
  {
    "idx": 15,
    "key": "item-15",
    "offsetTop": 750,
    "virtualHeight": 50,
  },
  {
    "idx": 16,
    "key": "item-16",
    "offsetTop": 800,
    "virtualHeight": 50,
  },
  {
    "idx": 17,
    "key": "item-17",
    "offsetTop": 850,
    "virtualHeight": 50,
  },
  {
    "idx": 18,
    "key": "item-18",
    "offsetTop": 900,
    "virtualHeight": 50,
  },
  {
    "idx": 19,
    "key": "item-19",
    "offsetTop": 950,
    "virtualHeight": 50,
  },
  {
    "idx": 20,
    "key": "item-20",
    "offsetTop": 1000,
    "virtualHeight": 50,
  },
  {
    "idx": 21,
    "key": "item-21",
    "offsetTop": 1050,
    "virtualHeight": 50,
  },
  {
    "idx": 22,
    "key": "item-22",
    "offsetTop": 1100,
    "virtualHeight": 50,
  },
  {
    "idx": 23,
    "key": "item-23",
    "offsetTop": 1150,
    "virtualHeight": 50,
  },
  {
    "idx": 24,
    "key": "item-24",
    "offsetTop": 1200,
    "virtualHeight": 50,
  },
  {
    "idx": 25,
    "key": "item-25",
    "offsetTop": 1250,
    "virtualHeight": 50,
  },
  {
    "idx": 26,
    "key": "item-26",
    "offsetTop": 1300,
    "virtualHeight": 50,
  },
  {
    "idx": 27,
    "key": "item-27",
    "offsetTop": 1350,
    "virtualHeight": 50,
  },
  {
    "idx": 28,
    "key": "item-28",
    "offsetTop": 1400,
    "virtualHeight": 50,
  },
]
`;
